### ISSUE TITLE
Error while fulfilling of JS: Request BrowserId : 1 not found it's likely the browser is already closed

### ISSUE THREAD

**Issue by maksimrudn** [cefsharp/CefSharp#4621]
### Is there an existing issue for this?

- [X] I have searched both open/closed issues, no issue already exists.

### CefSharp Version

117.2.40

### Operating System

Windows 10

### Architecture

x64

### .Net Version

.net 6.0

### Implementation

OffScreen

### Reproduction Steps

Version: Above 108

Steps:
Create instance of CefSharp.OffScreen.ChromiumWebBrowser
br = CefSharp.OffScreen.ChromiumWebBrowser();

Go by url 2 times:
br.Load(url)
await br.EvaluateScriptAsPromiseAsync(script);
br.Load(url)
await br.GetSourceAsync();
br.Load(url)

Fulfill JS
_browser.EvaluateScriptAsPromiseAsync(script)

### Expected behavior

Fulfilling of JS

### Actual behavior

JS Error Request BrowserId : 1 not found it's likely the browser is already closed

### Regression?

After regression to version 108 problem dissapeared.

### Known Workarounds

Likely problem is here
https://github.com/cefsharp/CefSharp/blob/master/CefSharp.BrowserSubprocess.Core/CefAppUnmanagedWrapper.cpp
Look at history..

### Does this problem also occur in the CEF Sample Application

Yes using WPF/OffScreen command line args

**Comment by amaitland** [comment]
> ### Reproduction Steps
> Version: Above 108

Please create a Unit test that reliably reproduces the problem and submit a `PR`.

https://github.com/cefsharp/CefSharp/tree/master/CefSharp.Test

Your code is incomplete and not enough for me to investigate this.

> Likely problem is here
> https://github.com/cefsharp/CefSharp/blob/master/CefSharp.BrowserSubprocess.Core/CefAppUnmanagedWrapper.cpp
> Look at history..

What is the problem **exactly**? Please provide details of your analysis.

Just saying look at history is meaningless. If you can see a specific error that needs fixing then submit a `PR`.

Likely this is the result of `upstream` changes to `CEF/Chromium`.

**Comment by maksimrudn** [comment]
> > ### Reproduction Steps
> > Version: Above 108
> 
> Please create a Unit test that reliably reproduces the problem and submit a `PR`.
> 
> https://github.com/cefsharp/CefSharp/tree/master/CefSharp.Test
> 
> Your code is incomplete and not enough for me to investigate this.
> 
> > Likely problem is here
> > https://github.com/cefsharp/CefSharp/blob/master/CefSharp.BrowserSubprocess.Core/CefAppUnmanagedWrapper.cpp
> > Look at history..
> 
> What is the problem **exactly**? Please provide details of your analysis.
> 
> Just saying look at history is meaningless. If you can see a specific error that needs fixing then submit a `PR`.
> 
> Likely this is the result of `upstream` changes to `CEF/Chromium`.

bro, are you interesting in solution of your project?
If you have concrete questions - ask me. I don't see any concrete questions. Looks like you just don't want to solve the problem.

**Comment by amaitland** [comment]
> Likely problem is here
>https://github.com/cefsharp/CefSharp/blob/master/CefSharp.BrowserSubprocess.Core/CefAppUnmanagedWrapper.cpp
> Look at history..

What is the problem exactly? Please provide details of your analysis.

I don't see a problem, so I need more information.

> You see last call of _browser.EvaluateScriptAsPromiseAsync(script) entails error.

I need a reliable means to reproduce the problem. The best way to achieve that is to provide a failing test case. 


> bro, are you interesting in solution of error of your project?

There's a high change this is a change in `CEF/Chromium` and will need to be addressed `upstream`.

> I don't see any concrete questions. Looks like you just don't want to solve the problem.

I'm very busy. If you want my assistance then you need to provide sufficient information. Just saying look at history is entirely pointless. Link to the exact commit that you think introduced the problem and provide an analysis of where you think the problem is.

**Comment by amaitland** [comment]
https://github.com/cefsharp/CefSharp/blob/bb30e237951c50707f9c75f89ccfb1c4dbadeff7/CefSharp.Test/Javascript/EvaluateScriptAsyncGenericTests.cs#L137

There are a number of unit tests that could be used as a base for reproducing this issue. Above is one such example.

If someone can create a test that reliability reproduces the issue, please create a pull request. Then I can have a look to see what's going on.

**Comment by daveime68** [comment]
This is still an issue as of 119.4.30 Core.

Once a web page has been loaded, and any automated action instigated on that page, for example clicking on a link, clicking a submit button, in fact ANYTHING that causes a new page to load in it's place, then any subsequent EvaluateScriptAsPromiseAsync or EvaluateScriptAsync executed on the new page will fail with the error message "Request BrowserId : 1 not found it's likely the browser is already closed.

As others have pointed out, the issue seems to be here:

https://github.com/cefsharp/CefSharp/blob/master/CefSharp.BrowserSubprocess.Core/CefAppUnmanagedWrapper.cpp

function OnProcessMessageReceived

auto browserWrapper = FindBrowserWrapper(browser->GetIdentifier());

is returning a nullptr, which triggers the error message.

But if you prefer to just respond with snarky comments or complain that no one will make a unit test for you, when the issue has been clearly described by multiple users, then honestly that's a you problem.

I'd imagine most people just pull packages via NuGet, and when they identify a problem and report it, it's on you as the maintainer to address it. No one else is obliged to make unit tests for you, or delve into the inner workings of C# wrappers to a C++ interface to a 3rd party chromium library.

Until then, your project will remain fundamentally broken.

**Comment by ficho-hun** [comment]
It seems that happening with only some Websites. 
I enabled the CEF logging, maybe can you find out something.

[
[cef.log](https://github.com/cefsharp/CefSharp/files/13874676/cef.log)
](url)

**Comment by netmaxik** [comment]
good day @amaitland 
I did https://github.com/netmaxik/CefSharp.MinimalExample.netcore special.
Use WPF project and press Test button. On title you will see Elements under mouse. Press Test again to stop. After you go to :   https://accounts.google.com/?hl=en, wait to load and press test again you see that error appears .

**Comment by amaitland** [comment]
> There's a high change this is a change in `CEF/Chromium` and will need to be addressed `upstream`.

As above.

> If someone can create a test that reliability reproduces the issue, please create a pull request. Then I can have a look to see what's going on.

When someone creates a **unit test** then I'll have another look.

Your asking for my time. I'm asking for for the community to contribute their time. It's a simple ask.

---

**MAKE SURE** you are waiting for the page to load correctly before executing javascript.

> br.Load(url)
> await br.EvaluateScriptAsPromiseAsync(script);

Calling `Load` then immediately evaluating javascript isn't reliable. You never know which V8 context the javascript will be executed in. If you perform a cross origin navigation then it's likely even more problematic.

> await br.LoadAsync(url)
> await br.EvaluateScriptAsPromiseAsync(script);

Waits for the page to load.

---

You can try disabling [site isolation trivls](https://peter.sh/experiments/chromium-command-line-switches/#disable-site-isolation-trials) to see if the behaviour changes.

```c#
settings.CefCommandLineArgs.Add("disable-site-isolation-trials");
```

**Comment by nebojsasandin** [comment]
@amaitland Thank you for your reply.

"disable-site-isolation-trials" did not help.

Here's the simplest test I could compile: https://github.com/nebojsasandin/ceftest

And it isn't much more complicated than few lines, so it can even fit into this comment here:

```
private static void RunCefTest()
{
    using (var browser = new ChromiumWebBrowser("https://www.google.com/"))
    {
        browser.WaitForInitialLoadAsync().Wait();
        browser.Load("https://accounts.google.com/");
        browser.WaitForNavigationAsync().Wait();
        var rez = browser.GetMainFrame().EvaluateScriptAsync<string>("document.querySelector(\"button[aria-haspopup='menu']\").innerText").Result;
    }
}
```
It is expected from this script to return "Create Account" text from the querySelector button, but what happens is the crash with the _Request BrowserId : 1 not found it's likely the browser is already closed_ inner exception.

**Comment by amaitland** [comment]
@nebojsasandin Thanks. You can create a test which mimics your example in https://github.com/cefsharp/CefSharp/blob/bb30e237951c50707f9c75f89ccfb1c4dbadeff7/CefSharp.Test/OffScreen/OffScreenBrowserTests.cs#L31

Please create a unit test inside the `CefSharp.Test` project and submit a `PR`. 

---

> If the original url is: about:blank, https://accounts.google.com, https://www.yahoo.com, https://www.bing.com, etc.... it all works fine.
> The point where it breaks is if the original url is: https://www.google.com, and then it navigates to https://accounts.google.com/
> This is why it was so hard to catch it. I haven't found another instance of this problem. Only when Google is being used.
> Also, Script Execution, without any return values, works just fine, and produces no exceptions.

What you are describing sounds like an `upstream` issue. 

---

You are using the `async` API in a `sync` fashion which is not supported.

There's an `async` example available at https://github.com/cefsharp/CefSharp.MinimalExample/blob/master/CefSharp.MinimalExample.OffScreen/Program.cs#L48 which you can use as a reference.

**Comment by nebojsasandin** [comment]
Not sure if this was what I was supposed to do (I've never done this before), but I forked the entire master branch, changed the offscreen test cs file, included the part which causes an exception, and initialized pull request on this one single file from my fork: https://github.com/cefsharp/CefSharp/pull/4741

I could not get it to compile or run though, there was a whole bunch of errors in the C++ sources which I'm unfamiliar with, so I could not see if the test truly replicates the issue described here, but I did test it in async fashion as well in the original test I sent you, and it did produce the same exception, so hopefully you'll be able to replicate the issue on your end and see what's causing  it in the first place, so we can get this resolved.

Thank you.

**Comment by amaitland** [comment]
> initialized pull request on this one single file from my fork: #4741

For the example provided by @nebojsasandin you can forcibly isolate the origins and the script executes correctly.

```c#
settings.CefCommandLineArgs.Add("isolate-origins", "https://www.google.com,https://accounts.google.com/");
```

**Comment by amaitland** [comment]
Loading `chrome://process-internals/#site-isolation` in `Chrome` I see 

![image](https://github.com/cefsharp/CefSharp/assets/307872/94f8d07a-5b5e-42d5-b62f-f55f2a05bd9e)

In the `CEF Sample Application (cefclient)` it looks like

![image](https://github.com/cefsharp/CefSharp/assets/307872/11fd22d2-890b-47a6-8e4a-7b7ffd317f0e)

So CEF doesn't have the same google urls isolated like `Chrome` does.

You can add the following command line arg to mimic the Chrome behaviour.

```c#
settings.CefCommandLineArgs.Add("isolate-origins", "https://accounts.google.com,https://chrome.google.com,https://chromewebstore.google.com");
```

**Comment by amaitland** [comment]
I've created https://github.com/chromiumembedded/cef/issues/3672 to see if `CEF` can set the `built-in` isolations by default.

Until then you can use the workaround I've listed above.

> It seems that happening with only some Websites.

@ficho-hun Did you observe the issue on websites other than the google ones?

> After you go to : https://accounts.google.com/?hl=en,

@netmaxik  If you are still experiencing this issue then please test with the workaround provided in https://github.com/cefsharp/CefSharp/issues/4621#issuecomment-2016708655

**Comment by netmaxik** [comment]
@amaitland I did test with : settings.CefCommandLineArgs.Add("isolate-origins", "https://accounts.google.com,https://chrome.google.com,https://chromewebstore.google.com");.

What I found out is :
1. If I start application - press Test button - then press Sign In or went to https://accounts.google.com/?hl=en, - I got problem. same error.  BUT
2. If I start application - DO not press Test and press Sign In to google or https://accounts.google.com/?hl=en - then press test -  its working normally.

**Comment by amaitland** [comment]
@netmaxik If you create a failing `Unit` test and create a **PR** then I will have a look. You can use https://github.com/cefsharp/CefSharp/pull/4741 as a reference for creating a test.

**Comment by nebojsasandin** [comment]
@netmaxik Have you tried adding more domains to the list? In my use-case I had to add few extra domains:
https://accounts.google.com,https://chrome.google.com,https://chromewebstore.google.com,https://reportcontent.google.com,https://support.google.com,https://www.google.com

**Comment by netmaxik** [comment]
@nebojsasandin  I did most off them.. nothing help. I do not know how to create test units... so do not know my next steps. But issue still continue :(

**Comment by nebojsasandin** [comment]
@netmaxik , when I test your code based on the instructions you provided (click on Test button, then navigate to Sign In, it still works when I include "https://www.google.com" to the isolate-origins).
...However, I realized that there are still issues with this even with the workaround. So maybe that's what you're talking about.
@amaitland , I've created a new pull request: https://github.com/cefsharp/CefSharp/pull/4771 with the new method created, the one which demonstrates that "isolate-origins" which includes all domains being visited is still producing the same error.

**Comment by amaitland** [comment]
> it seems that the CEF staff don't recognize this as a bug, or anything that needs to be changed, however, this does break the functionality of the CefSharp. Is there a more permanent or future-proof solution other than knowing which domains are breaking this entire facility?

The problem is in the `Alloy Bootstrap`, not something that can be fixed directly in `CefSharp`. The interesting news is that `CEF` is just started to remove the `Alloy Boostrap` (https://github.com/chromiumembedded/cef/issues/3685) and has added support for using the `Chrome Bootstrap` with `Alloy Style OSR` windows.

Starting in `M125` you'll be able to use `CefSettings.ChromeRuntime = true` and the same isolation settings a `Chrome` will be used.

I've added both tests (they'll only run locally, not on `Appveyor`) and migrated the `CefSharp.Test` project to using the `ChromeRuntime`. Both now pass without any changes or custom origin isolations.

In `~M127` the `Alloy Bootstrap` will be removed. 

> I've created a new pull request: #4771 with the new method created, the one which demonstrates that "isolate-origins" which includes all domains being visited is still producing the same error.

@nebojsasandin Thanks for providing the tests 👍 The problem is likely in the `alloy bootstrap` and should be resolved by switching to the `Chrome bootstrap`.

**Comment by heart-of-hero** [comment]
The above-mentioned issue still persists. Does anyone have a solution?

* Project type: WinForm, VB .NET Framework 4.8.1
* Operating System: Win 10, x64
* CEFSharp version: Stable releases 125 and 126 from NuGet.
* Other settings and prerequisites:
  - ChromeRuntime=true.
  - The error occurs regardless of whether "custom origin isolations" are enabled or not.
  - The timing for allowing Script execution has been confirmed through methods such as CanExecuteJavascriptInMainFrame.

* Issue occurrence process:
The issue occurs only after a page navigation is triggered by either manual or script operations on any webpage.
ExecuteScriptAsync() always runs normally, but EvaluateScriptAsync() throws the above error.
On the JavaScript side, CefSharp.BindObjectAsync() fails.

I am unsure whether this is a default behavior change in CEFSharp or a bug.
Looking forward to a solution, thank you.

**Comment by stormcleric** [comment]
True, I'm randomly getting this error after upgrading to latest version

**Comment by stormcleric** [comment]
> The above-mentioned issue still persists. Does anyone have a solution?
> 
> * Project type: WinForm, VB .NET Framework 4.8.1
> * Operating System: Win 10, x64
> * CEFSharp version: Stable releases 125 and 126 from NuGet.
> * Other settings and prerequisites:
>   
>   * ChromeRuntime=true.
>   * The error occurs regardless of whether "custom origin isolations" are enabled or not.
>   * The timing for allowing Script execution has been confirmed through methods such as CanExecuteJavascriptInMainFrame.
> * Issue occurrence process:
>   The issue occurs only after a page navigation is triggered by either manual or script operations on any webpage.
>   ExecuteScriptAsync() always runs normally, but EvaluateScriptAsync() throws the above error.
>   On the JavaScript side, CefSharp.BindObjectAsync() fails.
> 
> I am unsure whether this is a default behavior change in CEFSharp or a bug. Looking forward to a solution, thank you.

Well it seems to be fixed if you specifically set ChromeRuntime=false in your settings. At least this worked for me after upgrading, and everything seems to be working alright.

**Comment by GrabzIt** [comment]
I have also been experiencing the same issue with ChromeRuntime=true on random websites. It's not just related to Google websites.

I found it on: https://www.visualcom.pl/kontakt/wymagania-plikow-produkcyjnych'

**Comment by amaitland** [comment]
> Well it seems to be fixed if you specifically set ChromeRuntime=false in your settings. At least this worked for me after upgrading, and everything seems to be working alright.

If the problem only occurs when using the `Chrome Runtime` then an upstream issue will need to be raised. 

https://github.com/chromiumembedded/cef/issues

The alloy bootstrap will be removed in `M128`, you'll only be able to revert back for a short time.

> I have also been experiencing the same issue with ChromeRuntime=true on random websites. It's not just related to Google websites.
> 
> I found it on: https://www.visualcom.pl/kontakt/wymagania-plikow-produkcyjnych'

@GrabzIt Can you please create a unit test that reproduces the problem you are seeing (and submit a PR).

There are already tests at https://github.com/cefsharp/CefSharp/blob/63cce27846e2c112bd6bfa510db25a619d40e739/CefSharp.Test/Issues/Issue4621.cs#L14 you can use those as a basis, just add another to the same class

---

To set expectations, I've not seen anything here that suggests this is an issue that can be fixed directly in `CefSharp`. If `upstream` changes are required then this will likely take some time to get fixed as reproducing the issue directly in `C++` will be difficult.

I potential workaround would be to switch to using `Chrome DevTools Protocol` for executing javascript.

https://github.com/cefsharp/CefSharp.Dom?tab=readme-ov-file#dom-access

---

Hopefully at some point CEF will add equivalent functionality directly (https://github.com/chromiumembedded/cef/issues/3631)

**Comment by GrabzIt** [comment]
I have found that the errors happen after a long period of time, so hours or days, then carry on happening. So this isn't something that is suitable for a unit test.

I am going to try and use the Chrome DevTools Protocol to see if that fixes the issue.

**Comment by stormcleric** [comment]
> I have found that the errors happen after a long period of time, so hours or days, then carry on happening. So this isn't something that is suitable for a unit test.
> 
> I am going to try and use the Chrome DevTools Protocol to see if that fixes the issue.

Not on my specific case - they were after 10 solid minutes - but it's funny it's time dependant.

**Comment by drRobertDev** [comment]
> > I have found that the errors happen after a long period of time, so hours or days, then carry on happening. So this isn't something that is suitable for a unit test.
> > I am going to try and use the Chrome DevTools Protocol to see if that fixes the issue.
> 
> Not on my specific case - they were after 10 solid minutes - but it's funny it's time dependant.

yes also to me afte 10 minutes, dont know why this timing. (rendering working, on paint work, just JS code dont work)

**Comment by nabkawe** [comment]
> > I have found that the errors happen after a long period of time, so hours or days, then carry on happening. So this isn't something that is suitable for a unit test.
> > I am going to try and use the Chrome DevTools Protocol to see if that fixes the issue.
> 
> Not on my specific case - they were after 10 solid minutes - but it's funny it's time dependant.

I feel this is Google's way of breaking Chrome Extensions that need (JS level magic to do stuff like hide ads.)
I have the exact issue after upgrading from 112 to the latest stable nuget.

**Comment by nabkawe** [comment]
I had this error in another site, I "possibly" managed to fix it by turning off chromeruntime  from the CefSharp Settings 

`                CefSharp.WinForms.CefSettings se = new CefSharp.WinForms.CefSettings();
                se.CachePath = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData) + @"\CEF";
                se.EnablePrintPreview();
                se.CefCommandLineArgs.Add("enable-media-stream");
                se.JavascriptFlags = "--enable-media-stream";
                se.JavascriptFlags = "--AllowRunningInsecureContent";
                se.CefCommandLineArgs.Add("enable-usermedia-screen-capturing");
                se.CefCommandLineArgs.Add("disable-site-isolation-trials"); // didn't work 
                se.ChromeRuntime = false; // worked instantly.
                CefSharp.Cef.Initialize(se);
`

**Comment by onlylau** [comment]
Now they remove this "se.ChromeRuntime = false;" and i cant run my aplication! :\ start getting that error "Message = "Request BrowserId : 1 not found it's likely the browser is already closed""

**Comment by nuno84** [comment]
> Now they remove this "se.ChromeRuntime = false;" and i cant run my aplication! :\ start getting that error "Message = "Request BrowserId : 1 not found it's likely the browser is already closed""

I also have this issue. Did you find any workaround?

**Comment by onlylau** [comment]
no :\

Nuno ***@***.***> escreveu (quinta, 3/10/2024 à(s) 13:37):

**Comment by amaitland** [comment]
Please create a unit test inside the `CefSharp.Test` project and submit a `PR`. 

There are already tests at https://github.com/cefsharp/CefSharp/blob/63cce27846e2c112bd6bfa510db25a619d40e739/CefSharp.Test/Issues/Issue4621.cs#L14 you can use those as a basis, just add another to the same class

Until there's a **unit test** that reproduces the problem there's not really anything actionable here.

**Comment by aimcosoftware** [comment]
This is still an issue in 130.1.90.0 only when using `EvaluateScriptAsync`. `ExecuteScriptAsync `doesn't fail.
Not had this issue until Chrome bootstrap, all Alloy versions since 106 have worked.

**Steps**
Navigate to a page, on FrameLoadEnd, inject some DOM altering script , I'm using:

`for(elm of document.body.getElementsByTagName("div")){
	elm.style.border = "1px solid red"
}
`
Navigate to other pages, within 4-6 pages, you'll get the error.

CEFSharp is entirely broken with this change, happy to provide any assistance to track it down.

**Comment by nuno84** [comment]
But from what I understand, ExecuteScriptAsync doesn't return any data, and I need to get data from the browser. If I am right, than I cannot use that alternative.
Thank you for any suggestions.

**Comment by aimcosoftware** [comment]
@nuno84 `ExecuteScriptAsync `doesn't return any data. I only mentioned it because works, where `EvaluateScriptAsync `does not.
It also seems that the size of the page makes a difference, I'm testing with [these pages](https://aimcosoftware.co.uk/adev/CEFSharp/page1.html), they often fail on the 1st or 2nd navigate.

**Comment by amaitland** [comment]
> The page size seems to make a difference, I'm testing with [these pages](https://aimcosoftware.co.uk/adev/CEFSharp/page1.html), they often fail on the 1st or 2nd navigate.

@aimcosoftware Can you turn that into a `XUnit` test as per https://github.com/cefsharp/CefSharp/issues/4621#issuecomment-2393172786

I need a reliable way to reproduce the problem.

**Comment by mahelm** [comment]
I am experiencing this issue as well and I though I'd share my context here in case it could be of any help in finding a way to fix or mitigate the problem. In my case I run `EvaluateScriptAsync` on `IChromiumWebBrowserBase.FrameLoadEnd`. 
To reproduce the issue I have use two Angular applications, A and B, which are hosted on http://localhost:4200 and http://localhost:4201. 

- Load app A in the ChromiumWebBrowser => EvaluateScriptAsync success on site A
- Navigate to app B via a link in app A => EvaluateScriptAsync success on site B
- Navigate to app A via a link in app B => EvaluateScriptAsync failure with "Request BrowserId : 1 not found it's likely the browser is already closed"

I have tried with `settings.CefCommandLineArgs.Add("isolate-origins", "http://localhost,https://accounts.google.com,https://chrome.google.com,https://chromewebstore.google.com");` but without any luck.

Version:
CefSharp.WinForms - 129.0.110

**Comment by amaitland** [comment]
As above, I need more than just a list of steps to reproduce the problem, ideally a unit test, a fork of the minimal example would be ok I guess (would need to be turned into a unit test).

**Comment by mahelm** [comment]
> As above, I need more than just a list of steps to reproduce the problem, ideally a unit test, a fork of the minimal example would be ok I guess (would need to be turned into a unit test).

The key seems to be that you need to have two web applications served from the same "site" (where site is the scheme and registered domain name, including the public suffix, but ignoring subdomains, port, or path). These could probably be simple html-pages with links to navigate between the apps. The CEFSharp application should just run `EvaluateScriptAsync` when a page is loaded. I don't know how to set up this as a unit test the code base unfortunately. I hope this helps.

**Comment by mitchcapper** [comment]
Here is a pretty minimal example.  Originally I used a totally non google set of urls that loaded more resources and would crash in 3 loads, but even just using some google.com pages it can happen for me at least in 5 ever time.


The code in short:
```csharp
public MainWindow() {
	InitializeComponent();
	Loaded += MainWindow_Loaded;
}
private ChromiumWebBrowser browser;

private string theUrl = "https://google.com";
private string theUrl2 = "https://google.com/404";
private string theUrl3 = "https://google.com/503";
private async void MainWindow_Loaded(object sender, RoutedEventArgs e) {
	CefSharpSettings.RuntimeStyle = CefRuntimeStyle.Chrome;
	CefSettings settings = new() { CachePath = null, RootCachePath = @"x:\sandbox\test\trash_cache2", };
	Cef.Initialize(settings);
	browser = new ChromiumWebBrowser();
	contentCntrl.Content = browser;
	await Task.Delay(1000);

	await browser.NavigateTo(theUrl);
	var res = await browser.EvaluateScriptAsync<string>("5*6;");
	Log(res);
	await browser.NavigateTo(theUrl2);
	res = await browser.EvaluateScriptAsync<string>("5*6;");
	Log(res);
	
	await browser.NavigateTo(theUrl3);
	res = await browser.EvaluateScriptAsync<string>("5*6;");//original site generally throws here 
	Log(res);
	await browser.NavigateTo(theUrl2);
	res = await browser.EvaluateScriptAsync<string>("5*6;");
	Log(res);
	await browser.NavigateTo(theUrl);
	res = await browser.EvaluateScriptAsync<string>("5*6;");//for google sites throws here
	Log(res);
	await browser.NavigateTo(theUrl2);
	res = await browser.EvaluateScriptAsync<string>("5*6;");
	Log(res);

}

private void Log(string res) {
	Debug.WriteLine(res);
}
	}
	public static class helpers {
public static Task NavigateTo(this ChromiumWebBrowser browser, String url) {
	browser.Address = url;
	return browser.WaitForNavigationAsync(TimeSpan.FromSeconds(30));
}
```

Full project: 
[cefEvalIssueBare.zip](https://github.com/user-attachments/files/17915442/cefEvalIssueBare.zip)

**Comment by mitchcapper** [comment]
It also doesn't need 3 urls just navigating from 1 to 2 and back to 1 generally works too.

**Comment by amaitland** [comment]
Interestingly the `NavigateBetweenTwoUrlsBreaksJs` test passes if we disable the `BF Cache`.  

```c#
settings.CefCommandLineArgs.Add("disable-back-forward-cache");
```

This is a very specific case, so not sure if that's going to work in other instances. It **might** be worth trying for some of you.

**Comment by mahelm** [comment]
> Interestingly the `NavigateBetweenTwoUrlsBreaksJs` test passes if we disable the `BF Cache`.
> 
> ```cs
> settings.CefCommandLineArgs.Add("disable-back-forward-cache");
> ```
> 
> This is a very specific case, so not sure if that's going to work in other instances. It **might** be worth trying for some of you.

Thanks! This fix my our issue at least!

**Comment by SASoper** [comment]
Wonderful! This has fixed this issue for me as well!

**Comment by ficho-hun** [comment]
For me, the error still exists. I load the URL only once, but the website's JavaScript script performs additional navigations.

**Comment by mitchcapper** [comment]
@ficho-hun if you can put together a test case like amaitland's comment https://github.com/cefsharp/CefSharp/issues/4621#issuecomment-2500082429

likely to be responded far faster, especially now that many of us can't repro with that flag set.

**Comment by onlylau** [comment]
One of the things I noticed, is that I lose the debug of my javascript code, 

Ex: Console.log("Finished")

When running the javascript in Visuastudio, nothing appears in the output.

**Comment by mitchcapper** [comment]
> One of the things I noticed, is that I lose the debug of my javascript code, 
> 
> Ex: Console.log("Finished")
> 
> When running the javascript in Visuastudio, nothing appears in the output.

Under what conditions?  Can you put a repro case together like above?

**Comment by netmaxik** [comment]
Good day
same problem in https://www.quora.com/ site. :)

Regards
Maxim

**Comment by pop1989bb** [comment]
I unfortunately experience this problem since 2 weeks. Updated to 135, but still unsolved. What's the current state?

**Comment by mitchcapper** [comment]
@pop1989bb I believe this is an upstream issue.  I think  it is related to: https://github.com/chromiumembedded/cef/issues/3631 around difficulty in getting results from JS calls  that amaitland referenced above.  Disabling the back/forward cache resolves this, at the cost of obviously losing that feature.  One item of note, it seems to only happen when going to the same url multiple times one might also be able to get around this by  appending a cache break to the url as part of a query string without having to disable the cache entirely.

**Comment by mitchcapper** [comment]
@pop1989bb for fastest resolution please put together a test case that works on master like: https://github.com/cefsharp/CefSharp/issues/4621#issuecomment-2500082429 that fails when bf cache is disabled (so add that to the init)

**Comment by pop1989bb** [comment]
@mitchcapper 
I tried to provide an example but it doesn't really work (internal stuff etc.).
But I can confirm that with the most current release the problem still is there.
And @netmaxik confirms it with another page so maybe this example is enough?

**Comment by mitchcapper** [comment]
@pop1989bb can you replicate it with the site @netmaxik provided? If so please provide a test case like above for fastest resolution.   I tried but was unable to.  Otherwise I would try to par down exactly what internally is able to cause it (is it only certain pages? How basic can the page be?)  then copy the source to an external resource and remove anything sensitive but still can repro the issue.

**Comment by pop1989bb** [comment]
@mitchcapper 
I will look for a basic example and try to provide it.

**Comment by mitchcapper** [comment]
Thanks @yevklim for some great alternate work around to disabling the BF cache. Here is where I believe this issue currently stands:

### Summary

This bug is likely due to an issue in CEF getting triggered by CefSharp's code that allows returning a result from a javascript execute call.  This is harder to reproduce in the native client given Cef does not actually natively support the behavior of returning said result.  1.5 years ago amaitland opened https://github.com/chromiumembedded/cef/issues/3631 requesting such a feature added to Cef.   If it was either this bug would go away or be far easier to reproduce in Cef.

### 4 known workarounds

- Avoid calling `EvaluateScriptAsync` or `bindObjectAsync` if you don't need them
- Disabling the back-forward-cache with: `settings.CefCommandLineArgs.Add("disable-back-forward-cache");` resolves the test case from https://github.com/cefsharp/CefSharp/blob/85cbb485d614a868c040a7f1adf5703a8b4f9118/CefSharp.Test/Issues/Issue4621.cs .  Some people say they still are able to get this crash even with the workaround in place but we don't have an easily reproducible case so looking into that possibility is much harder.  This is likely the easiest solution for most people, as long as you are OK without the back-forward cache for now.
- Using the devtools API for javascript execution ie: https://github.com/cefsharp/CefSharp.Dom?tab=readme-ov-file#evaluate-javascript devtools does not seem to trigger the same crash given the alternate mechanism in which it works.  The devtools DOM api has a slew of other useful features, and being devtools based makes it far less likely to break than CEF itself as the devtools protocol and interface are largely exactly what is in chrome so has first party support
- [yevklim's solution](https://github.com/cefsharp/CefSharp/issues/4621#issuecomment-2880349984) above of emulating the current behavior using JS callbacks and message passing.  For more complex uses one could build a wrapper on both sides to emulate most of the existing behavior if one of the above solutions didn't work.

### Ways to help
- If you still have this problem with the back-forward-cache disabled see if you can create a test case like: https://github.com/cefsharp/CefSharp/blob/85cbb485d614a868c040a7f1adf5703a8b4f9118/CefSharp.Test/Issues/Issue4621.cs reproducing the issue. It comes down to adding  `settings.CefCommandLineArgs.Add("disable-back-forward-cache");` at the start of the test and then making whatever changes to get it to crash on the sites you have found trigger it.  As with any JS calls make sure you are waiting for the page to fully load before executing javascript as well.   There are ways (especially with devtools) to execute javascript prior to any other javascript is run (if required).
- If you have found an alternate workaround to above known workarounds please post it
- If you are able to reproduce the issue with back forward cache enabled (ie the test case) in a simplified example with CEF base then that would be helpful to make it easier to file an upstream bug.  In theory one approach here could to strip down the CEFSharp code path for EvaluateScriptAsync to the minimum to trigger it.   I am sure CEF prefers bugs reproducible with cefclient but given it doesn't support
- [sponsor amaitland](https://github.com/sponsors/amaitland) he constantly is working to stay up with changes from CEF and improve this library, not only do monetary contributions let him spend more time on it but he is often able to prioritize issues that are affecting those helping keep this project viable

**Comment by mitchcapper** [comment]
> [@mitchcapper](https://github.com/mitchcapper) [@yevklim](https://github.com/yevklim) Unfortunately cefSharp.postMessage seems to have the same problem for me.

Are you actually getting the "Request BrowserId : 1 not found it's likely the browser is already closed" or do you just mean the post message does not work?

After a redirect does EvaluateMessage get fired?  I assume given you mention the handlers are still in the dictionary.   Just to confirm if you have a long running handler on JavascriptMessageReceived get called?  Are there any javascript errors from running cefSharp.postMessage? (ie devtools)

Is there a specific easy site to use the above code on to reproduce?

**Comment by pop1989bb** [comment]
Try it on Marktplaats (www.maarktplaats.nl). Maybe you can reproduce it.
To be clear: I am not sure what fails. I only can tell that they never were responded. I'll try the dev tools that's a good idea.

**Comment by yevklim** [comment]
Clearly, there are some very unfortunate bugs in either CEF or CefSharp at this point, which cause problems downstream. Besides the half-broken IPC that I described previously, my company encountered multiple app crashes caused by a null pointer somewhere in the Chromium source code.

Whilst I believe it would only be right to help the maintainers to resolve the issues, some people downstream simply don't have the time and resources for it, but do have the necessity of solving the problem sooner rather than later. Luckily, there are good alternatives, e.g., [**WebView2**](https://learn.microsoft.com/en-us/microsoft-edge/webview2/) from Microsoft. Some people may not like the fact that it's from Microsoft, but it is objectively more stable than CefSharp at this time.

My company was able to switch to WebView2 in a matter of days. It's not perfect, and it has its own issues, but nothing critical.

I'm writing this not to disrespect CefSharp, but to suggest a viable alternative for those who need it.
